src/groups/mqb/mqbstat/mqbstat_statcontroller.cpp

```diff
@@ -81,6 +82,20 @@ const bsls::Types::Int64 k_NS_PER_MESSAGE = 15 *
 
 const char k_PUBLISHINTERVAL_SUFFIX[] = ".PUBLISHINTERVAL";
 
+void portsDeleter(
+    bsl::unordered_map<bsl::string, bslma::ManagedPtr<mwcst::StatContext> >*
+                              map,
+    bslmt::Mutex*             mutex,
+    const mwcst::StatContext& context)
+{
+    // Lookup the port's StatContext and remove it from the 'map'
+    bslmt::LockGuard<bslmt::Mutex> guard(mutex);  // LOCK
```

Comment:
Unfortunately, the threadsafety of `StatContext` class is only partial.
https://github.com/bloomberg/blazingmq/blob/7d4b3d7caa075e5cca89995008e6fad30082b3f2/src/groups/mwc/mwcst/mwcst_statcontext.h#L96-L98
For example, I can't iterate over subcontexts that can be added/deleted without additional synchronisation anywhere except the `snapshot()` method or its synchronious pre- or post- steps, as handling the deleted subcontexts happens inside this method. Please find below some arguments in favor of the current imlementation:
Each channel stores its own `StatContext`, and this approach hasn't changed. What I have done is another 'layer' of `StatContext`s between channel contexts and their parent. This layer serves for breaking down the channels stats by ports. I really need to have only one `StatContext` for each port as I report aggregated stats. The inter-thread synchronisation is required here, because:
1. `portsDeleter()` is always called from `StatController`'s thread, but `TCPSessionFactory::channelStatContextCreator` and, hence,  constructors and destructors of channel-related `StatContext`s are called from `TCPSessionFactory`'s threads.
2. Although 'StatContext::addSubcontext()' is a thread-safe method, it will create a new `StatContext` instance even if there are sub-contexts with the same name.
So, in order to add a **channel** StatContext, I need to check if there is a **port** StatContext for that port and use it, otherways I have to create a new **port** `Statcontext`. After that I add a **channel** sub context to it. So I must have a **ManagedPtr**'s to the `StatContext`s of the **ports** and to lock a container (currently, the unordered_map), where they are stored. Otherways race conditions are possible, either with two insertions of the same name **port** `StatContext`s, or deletion during the adding a new **channel** subcontext.

If you have an idea how to reach it without the synchronization, please let me know

Reply:
> If you have an idea how to reach it without the synchronization, please let me know
@waldgange let me think about this for some time

Reply:
Hi @waldgange
I think the correct way to do it is to pass `ManagedPtr` to the `StatChannel` here, instead of `shared_ptr`:
https://github.com/bloomberg/blazingmq/blob/a6a6cefbb9b8cab1dc0b2bc0b1ee7c9bf3f8b05c/src/groups/mwc/mwcio/mwcio_statchannelfactory.cpp#L126-L133

So, `StatChannel` keeps ownership of this `StatContext`. When a `StatChannel` goes down, the `ManagedPtr` to its `StatContext` goes down too.

Note that it requires to change `MWC` interfaces. However, `StatChannel` and `StatChannelFactory` are used only within `BlazingMQ`, so I don't see any harm in this. Also, since we plan to own `ManagedPtr` in `StatChannel`, the `StatChannel` class is not copy constructible anymore, so we might explicitly delete copy constructor and assignment